File: go-backend/internal/service/image_generation_service.go
package service

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"strings"
	"time"

	"github.com/yupi/airouter/go-backend/internal/common"
	"github.com/yupi/airouter/go-backend/internal/errno"
	"github.com/yupi/airouter/go-backend/internal/model/dto"
	"github.com/yupi/airouter/go-backend/internal/model/entity"
	"github.com/yupi/airouter/go-backend/internal/repository"
)

const (
	defaultImageModel   = "qwen-image-plus"
	defaultImageSize    = "1024*1024"
	defaultImageCount   = 1
	imagePollInterval   = 2 * time.Second
	imagePollMaxRetries = 60
	imageTokensPerItem  = 1000
)

type ImageGenerationService struct {
	imageRecordRepo *repository.ImageGenerationRecordRepository
	modelService    *ModelService
	providerService *ProviderService
	userService     *UserService
	balanceService  *BalanceService
}

func NewImageGenerationService(
	imageRecordRepo *repository.ImageGenerationRecordRepository,
	modelService *ModelService,
	providerService *ProviderService,
	userService *UserService,
	balanceService *BalanceService,
) *ImageGenerationService {
	return &ImageGenerationService{
		imageRecordRepo: imageRecordRepo,
		modelService:    modelService,
		providerService: providerService,
		userService:     userService,
		balanceService:  balanceService,
	}
}

func (s *ImageGenerationService) GenerateImage(
	request dto.ImageGenerationRequest,
	userID *int64,
	apiKeyID *int64,
	clientIP string,
) (*dto.ImageGenerationResponse, error) {
	start := time.Now()
	prompt := strings.TrimSpace(request.Prompt)
	if prompt == "" {
		return nil, errno.NewWithMessage(errno.ParamsError, "提示词不能为空")
	}

	modelKey := strings.TrimSpace(request.Model)
	if modelKey == "" {
		modelKey = defaultImageModel
	}
	size := strings.TrimSpace(request.Size)
	if size == "" {
		size = defaultImageSize
	}
	n := defaultImageCount
	if request.N != nil && *request.N > 0 {
		n = *request.N
	}

	if userID != nil && *userID > 0 {
		disabled, disabledErr := s.userService.IsUserDisabled(*userID)
		if disabledErr != nil {
			return nil, disabledErr
		}
		if disabled {
			return nil, errno.NewWithMessage(errno.ForbiddenError, "账号已被禁用，无法使用服务")
		}
		quotaEnough, quotaErr := s.userService.CheckQuota(*userID)
		if quotaErr != nil {
			return nil, quotaErr
		}
		if !quotaEnough {
			return nil, errno.NewWithMessage(errno.OperationError, "Token配额已用尽，请联系管理员增加配额")
		}
	}

	model, err := s.modelService.GetByModelKey(modelKey)
	if err != nil {
		return nil, err
	}
	if model.ModelType != "image" {
		return nil, errno.NewWithMessage(errno.ParamsError, "模型不存在或不是绘图模型")
	}
	provider, err := s.providerService.GetProviderByID(model.ProviderID)
	if err != nil {
		return nil, err
	}

	estimatedCost := model.InputPrice * float64(n)
	if userID != nil && *userID > 0 {
		ok, checkErr := s.balanceService.CheckBalance(*userID, estimatedCost)
		if checkErr != nil {
			return nil, checkErr
		}
		if !ok {
			return nil, errno.NewWithMessage(errno.OperationError, fmt.Sprintf("账户余额不足，生成%d张图片预计需要¥%.4f，请先充值", n, estimatedCost))
		}
	}

	response, genErr := s.callImageModel(provider, modelKey, prompt, size, n, request.ResponseFormat)
	duration := int(time.Since(start).Milliseconds())
	if genErr != nil {
		if userID != nil && *userID > 0 {
			_ = s.imageRecordRepo.Create(&entity.ImageGenerationRecord{
				UserID:       *userID,
				APIKeyID:     apiKeyID,
				ModelID:      model.ID,
				ModelKey:     modelKey,
				Prompt:       prompt,
				Size:         size,
				Quality:      request.Quality,
				Status:       "failed",
				Cost:         0,
				Duration:     duration,
				ErrorMessage: genErr.Error(),
				ClientIP:     clientIP,
				CreateTime:   time.Now(),
			})
		}
		return nil, genErr
	}

	actualCount := len(response.Data)
	if actualCount == 0 {
		actualCount = n
	}
	actualCost := model.InputPrice * float64(actualCount)
	for _, item := range response.Data {
		record := &entity.ImageGenerationRecord{
			UserID:        zeroIfNil(userID),
			APIKeyID:      apiKeyID,
			ModelID:       model.ID,
			ModelKey:      modelKey,
			Prompt:        prompt,
			RevisedPrompt: item.RevisedPrompt,
			ImageURL:      item.URL,
			ImageData:     item.B64JSON,
			Size:          size,
			Quality:       request.Quality,
			Status:        "success",
			Cost:          model.InputPrice,
			Duration:      duration,
			ClientIP:      clientIP,
			CreateTime:    time.Now(),
		}
		_ = s.imageRecordRepo.Create(record)
	}

	if userID != nil && *userID > 0 {
		_ = s.userService.DeductTokens(*userID, int64(actualCount*imageTokensPerItem))
		if actualCost > 0 {
			desc := fmt.Sprintf("网页图片生成 - %s x%d", modelKey, actualCount)
			if apiKeyID != nil && *apiKeyID > 0 {
				desc = fmt.Sprintf("API图片生成 - %s x%d", modelKey, actualCount)
			}
			if deductErr := s.balanceService.DeductBalance(*userID, actualCost, nil, desc); deductErr != nil {
				log.Printf("image deduct balance failed: userId=%d model=%s cost=%.4f err=%v", *userID, modelKey, actualCost, deductErr)
				return nil, deductErr
			}
		}
	}

	return response, nil
}

func (s *ImageGenerationService) ListUserRecords(userID, pageNum, pageSize int64) (common.PageResponse[entity.ImageGenerationRecord], error) {
	if userID <= 0 {
		return common.PageResponse[entity.ImageGenerationRecord]{}, errno.New(errno.ParamsError)
	}
	if pageNum <= 0 {
		pageNum = 1
	}
	if pageSize <= 0 {
		pageSize = 10
	}
	return s.imageRecordRepo.ListByUserID(userID, pageNum, pageSize)
}

func (s *ImageGenerationService) callImageModel(
	provider *entity.ModelProvider,
	modelKey, prompt, size string,
	n int,
	responseFormat string,
) (*dto.ImageGenerationResponse, error) {
	if strings.Contains(provider.BaseURL, "dashscope.aliyuncs.com") && strings.HasPrefix(modelKey, "qwen-image") {
		return s.callQwenAsync(provider, modelKey, prompt, size, n)
	}
	return s.callOpenAIImage(provider, modelKey, prompt, size, n, responseFormat)
}

func (s *ImageGenerationService) callOpenAIImage(
	provider *entity.ModelProvider,
	modelKey, prompt, size string,
	n int,
	responseFormat string,
) (*dto.ImageGenerationResponse, error) {
	baseURL := strings.TrimRight(provider.BaseURL, "/")
	url := baseURL + "/v1/images/generations"
	body := map[string]any{
		"model":  modelKey,
		"prompt": prompt,
		"n":      n,
		"size":   size,
	}
	if strings.TrimSpace(responseFormat) != "" {
		body["response_format"] = responseFormat
	}
	payload, _ := json.Marshal(body)
	req, _ := http.NewRequest(http.MethodPost, url, bytes.NewBuffer(payload))
	req.Header.Set("Authorization", "Bearer "+provider.APIKey)
	req.Header.Set("Content-Type", "application/json")

	client := &http.Client{Timeout: 60 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return nil, errno.NewWithMessage(errno.SystemError, "调用图片生成API失败: "+err.Error())
	}
	defer resp.Body.Close()
	respBody, _ := io.ReadAll(resp.Body)
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, errno.NewWithMessage(errno.SystemError, "调用图片生成API失败: "+string(respBody))
	}
	var result struct {
		Created int64 `json:"created"`
		Data    []struct {
			URL           string `json:"url"`
			B64JSON       string `json:"b64_json"`
			RevisedPrompt string `json:"revised_prompt"`
		} `json:"data"`
	}
	if err = json.Unmarshal(respBody, &result); err != nil {
		return nil, errno.NewWithMessage(errno.SystemError, "解析图片生成响应失败")
	}
	items := make([]dto.ImageDataEntry, 0, len(result.Data))
	for _, item := range result.Data {
		items = append(items, dto.ImageDataEntry{
			URL:           item.URL,
			B64JSON:       item.B64JSON,
			RevisedPrompt: item.RevisedPrompt,
		})
	}
	return &dto.ImageGenerationResponse{
		Created: result.Created,
		Data:    items,
	}, nil
}

func (s *ImageGenerationService) callQwenAsync(
	provider *entity.ModelProvider,
	modelKey, prompt, size string,
	n int,
) (*dto.ImageGenerationResponse, error) {
	base := strings.Replace(strings.TrimRight(provider.BaseURL, "/"), "/compatible-mode", "", 1)
	createURL := base + "/api/v1/services/aigc/text2image/image-synthesis"
	createBody := map[string]any{
		"model": modelKey,
		"input": map[string]any{
			"prompt": prompt,
		},
		"parameters": map[string]any{
			"size": size,
			"n":    n,
		},
	}
	payload, _ := json.Marshal(createBody)
	req, _ := http.NewRequest(http.MethodPost, createURL, bytes.NewBuffer(payload))
	req.Header.Set("Authorization", "Bearer "+provider.APIKey)
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("X-DashScope-Async", "enable")

	client := &http.Client{Timeout: 60 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return nil, errno.NewWithMessage(errno.SystemError, "创建图片生成任务失败: "+err.Error())
	}
	defer resp.Body.Close()
	createRespBody, _ := io.ReadAll(resp.Body)
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, errno.NewWithMessage(errno.SystemError, "创建图片生成任务失败: "+string(createRespBody))
	}
	var createResp struct {
		Output struct {
			TaskID string `json:"task_id"`
		} `json:"output"`
		Message string `json:"message"`
	}
	if err = json.Unmarshal(createRespBody, &createResp); err != nil || createResp.Output.TaskID == "" {
		return nil, errno.NewWithMessage(errno.SystemError, "创建图片生成任务失败")
	}

	taskURL := base + "/api/v1/tasks/" + createResp.Output.TaskID
	for i := 0; i < imagePollMaxRetries; i++ {
		time.Sleep(imagePollInterval)
		taskReq, _ := http.NewRequest(http.MethodGet, taskURL, nil)
		taskReq.Header.Set("Authorization", "Bearer "+provider.APIKey)
		taskResp, taskErr := client.Do(taskReq)
		if taskErr != nil {
			continue
		}
		taskBody, _ := io.ReadAll(taskResp.Body)
		taskResp.Body.Close()
		var taskResult struct {
			Output struct {
				TaskStatus string `json:"task_status"`
				Message    string `json:"message"`
				Results    []struct {
					URL string `json:"url"`
				} `json:"results"`
			} `json:"output"`
		}
		if err = json.Unmarshal(taskBody, &taskResult); err != nil {
			continue
		}
		switch taskResult.Output.TaskStatus {
		case "SUCCEEDED":
			items := make([]dto.ImageDataEntry, 0, len(taskResult.Output.Results))
			for _, item := range taskResult.Output.Results {
				if item.URL != "" {
					items = append(items, dto.ImageDataEntry{
						URL:           item.URL,
						RevisedPrompt: prompt,
					})
				}
			}
			if len(items) == 0 {
				return nil, errno.NewWithMessage(errno.SystemError, "任务成功但未返回图片")
			}
			return &dto.ImageGenerationResponse{
				Created: time.Now().Unix(),
				Data:    items,
			}, nil
		case "FAILED":
			return nil, errno.NewWithMessage(errno.SystemError, "图片生成失败: "+taskResult.Output.Message)
		}
	}
	return nil, errno.NewWithMessage(errno.SystemError, "图片生成超时")
}

func zeroIfNil(v *int64) int64 {
	if v == nil {
		return 0
	}
	return *v
}

File: python-backend/app/services/image_generation_service.py
"""Image generation service."""

from __future__ import annotations

import asyncio
import logging
import time
from decimal import Decimal
from math import ceil

import httpx
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.constants import (
    DEFAULT_IMAGE_COUNT,
    DEFAULT_IMAGE_MODEL,
    DEFAULT_IMAGE_SIZE,
    DEFAULT_PAGE_NUM,
    DEFAULT_PAGE_SIZE,
    ErrorCode,
    IMAGE_RECORD_STATUS_FAILED,
    IMAGE_RECORD_STATUS_SUCCESS,
    IMAGE_TOKENS_PER_IMAGE,
    MAX_PAGE_SIZE,
    MODEL_TYPE_IMAGE,
)
from app.exceptions.business_exception import BusinessException
from app.models.image_generation_record import ImageGenerationRecord
from app.schemas.common import PageData
from app.schemas.image import (
    ImageData,
    ImageGenerationRecordVO,
    ImageGenerationRequest,
    ImageGenerationResponse,
)
from app.services.balance_service import BalanceService
from app.services.model_provider_service import ModelProviderService
from app.services.model_service import ModelService
from app.services.quota_service import QuotaService
from app.services.user_service import UserService

logger = logging.getLogger("app")


class ImageGenerationService:
    def __init__(self, db: AsyncSession) -> None:
        self.db = db
        self.model_service = ModelService(db)
        self.model_provider_service = ModelProviderService(db)
        self.quota_service = QuotaService(db)
        self.balance_service = BalanceService(db)
        self.user_service = UserService(db)

    async def generate_image(
        self,
        request: ImageGenerationRequest,
        user_id: int,
        api_key_id: int | None,
        client_ip: str | None,
    ) -> ImageGenerationResponse:
        start_ms = int(time.time() * 1000)
        if not request.prompt or not request.prompt.strip():
            raise BusinessException(ErrorCode.PARAMS_ERROR, "提示词不能为空")

        if await self.user_service.is_user_disabled(user_id):
            raise BusinessException(ErrorCode.FORBIDDEN_ERROR, "账号已被禁用，无法使用服务")
        if not await self.quota_service.check_quota(user_id):
            raise BusinessException(ErrorCode.OPERATION_ERROR, "Token配额已用尽，请联系管理员增加配额")

        model_key = request.model or DEFAULT_IMAGE_MODEL
        size = request.size or DEFAULT_IMAGE_SIZE
        image_count = DEFAULT_IMAGE_COUNT

        model = await self.model_service.get_by_model_key(model_key)
        if model is None or model.model_type != MODEL_TYPE_IMAGE:
            raise BusinessException(ErrorCode.PARAMS_ERROR, "模型不存在或不是绘图模型")
        provider = await self.model_provider_service.get_by_id(model.provider_id)
        if provider is None:
            raise BusinessException(ErrorCode.PARAMS_ERROR, "模型提供者不存在")

        estimated_cost = Decimal(model.input_price or 0) * Decimal(image_count)
        if not await self.balance_service.check_balance(user_id, estimated_cost):
            raise BusinessException(
                ErrorCode.OPERATION_ERROR,
                f"账户余额不足，生成{image_count}张图片预计需要¥{estimated_cost}，请先充值",
            )

        try:
            response = await self._call_dashscope_image_api(
                provider_base_url=provider.base_url,
                provider_api_key=provider.api_key,
                model_key=model.model_key,
                prompt=request.prompt,
                size=size,
                n=image_count,
            )
            duration = int(time.time() * 1000) - start_ms
            actual_count = len(response.data) if response.data else image_count
            actual_cost = Decimal(model.input_price or 0) * Decimal(actual_count)

            for item in response.data:
                self.db.add(
                    ImageGenerationRecord(
                        user_id=user_id,
                        api_key_id=api_key_id,
                        model_id=model.id,
                        model_key=model_key,
                        prompt=request.prompt,
                        revised_prompt=item.revised_prompt,
                        image_url=item.url,
                        image_data=item.b64_json,
                        size=size,
                        quality=request.quality,
                        status=IMAGE_RECORD_STATUS_SUCCESS,
                        cost=Decimal(model.input_price or 0),
                        duration=duration,
                        client_ip=client_ip,
                    )
                )
            await self.db.commit()

            await self.quota_service.deduct_tokens(user_id, actual_count * IMAGE_TOKENS_PER_IMAGE)
            if actual_cost > 0:
                source = "API图片生成" if api_key_id is not None else "网页图片生成"
                await self.balance_service.deduct_balance(
                    user_id,
                    actual_cost,
                    request_log_id=None,
                    description=f"{source} - {model_key} x{actual_count}",
                )

            logger.info("图片生成成功：用户 %s，模型 %s，数量 %s", user_id, model_key, actual_count)
            return response
        except BusinessException as exc:
            await self._save_failed_record(
                user_id=user_id,
                api_key_id=api_key_id,
                model_id=model.id,
                model_key=model_key,
                request=request,
                size=size,
                client_ip=client_ip,
                start_ms=start_ms,
                error_message=str(exc),
            )
            raise
        except Exception as exc:
            await self._save_failed_record(
                user_id=user_id,
                api_key_id=api_key_id,
                model_id=model.id,
                model_key=model_key,
                request=request,
                size=size,
                client_ip=client_ip,
                start_ms=start_ms,
                error_message=str(exc),
            )
            raise BusinessException(ErrorCode.SYSTEM_ERROR, f"图片生成失败: {exc}") from exc

    async def list_user_records(self, user_id: int, page_num: int, page_size: int) -> PageData[ImageGenerationRecordVO]:
        safe_page_num = page_num if page_num > 0 else DEFAULT_PAGE_NUM
        safe_page_size = page_size if page_size > 0 else DEFAULT_PAGE_SIZE
        safe_page_size = min(safe_page_size, MAX_PAGE_SIZE)

        base_stmt = select(ImageGenerationRecord).where(ImageGenerationRecord.user_id == user_id)
        total_row = await self.db.scalar(select(func.count()).select_from(base_stmt.subquery())) or 0
        rows = (
            await self.db.scalars(
                base_stmt.order_by(ImageGenerationRecord.create_time.desc())
                .offset((safe_page_num - 1) * safe_page_size)
                .limit(safe_page_size)
            )
        ).all()
        return PageData[ImageGenerationRecordVO](
            records=[ImageGenerationRecordVO.model_validate(item) for item in rows],
            pageNumber=safe_page_num,
            pageSize=safe_page_size,
            totalPage=ceil(total_row / safe_page_size) if total_row else 0,
            totalRow=total_row,
            optimizeCountQuery=True,
        )

    async def _save_failed_record(
        self,
        user_id: int,
        api_key_id: int | None,
        model_id: int,
        model_key: str,
        request: ImageGenerationRequest,
        size: str,
        client_ip: str | None,
        start_ms: int,
        error_message: str,
    ) -> None:
        duration = int(time.time() * 1000) - start_ms
        self.db.add(
            ImageGenerationRecord(
                user_id=user_id,
                api_key_id=api_key_id,
                model_id=model_id,
                model_key=model_key,
                prompt=request.prompt,
                size=size,
                quality=request.quality,
                status=IMAGE_RECORD_STATUS_FAILED,
                cost=Decimal("0"),
                duration=duration,
                error_message=error_message[:512],
                client_ip=client_ip,
            )
        )
        await self.db.commit()

    async def _call_dashscope_image_api(
        self,
        provider_base_url: str,
        provider_api_key: str,
        model_key: str,
        prompt: str,
        size: str,
        n: int,
    ) -> ImageGenerationResponse:
        normalized_base_url = provider_base_url.rstrip("/").replace("/compatible-mode", "")
        create_task_url = f"{normalized_base_url}/api/v1/services/aigc/text2image/image-synthesis"
        headers = {
            "Authorization": f"Bearer {provider_api_key}",
            "Content-Type": "application/json",
            "X-DashScope-Async": "enable",
        }
        payload = {
            "model": model_key,
            "input": {"prompt": prompt},
            "parameters": {"size": size, "n": n},
        }

        async with httpx.AsyncClient(timeout=30) as client:
            create_resp = await client.post(create_task_url, headers=headers, json=payload)
            create_data = create_resp.json()

            task_id = (((create_data.get("output") or {}).get("task_id")) or "").strip()
            if not task_id:
                message = create_data.get("message") or "未返回任务ID"
                raise BusinessException(ErrorCode.SYSTEM_ERROR, f"创建图片生成任务失败：{message}")

            task_url = f"{normalized_base_url}/api/v1/tasks/{task_id}"
            max_retries = 60
            for _ in range(max_retries):
                await asyncio.sleep(2)
                task_resp = await client.get(task_url, headers={"Authorization": f"Bearer {provider_api_key}"}, timeout=10)
                task_data = task_resp.json()
                output = task_data.get("output") or {}
                task_status = output.get("task_status")
                if task_status == "SUCCEEDED":
                    results = output.get("results") or []
                    image_data = [
                        ImageData(url=item.get("url"), revisedPrompt=prompt)
                        for item in results
                        if item.get("url")
                    ]
                    if not image_data:
                        raise BusinessException(ErrorCode.SYSTEM_ERROR, "任务成功但未返回图片")
                    return ImageGenerationResponse(created=int(time.time()), data=image_data)
                if task_status == "FAILED":
                    message = output.get("message") or "任务失败"
                    raise BusinessException(ErrorCode.SYSTEM_ERROR, f"图片生成失败：{message}")

            raise BusinessException(ErrorCode.SYSTEM_ERROR, "图片生成超时")
